pkg/querier/parquet_queryable.go

```diff
@@ -0,0 +1,346 @@
+
+	if err != nil {
+		return nil, err
+	}
+	manager, err := services.NewManager(blockStorageQueryable)
+	if err != nil {
+		return nil, err
+	}
+
+	cDecoder := schema.NewPrometheusParquetChunksDecoder(chunkenc.NewPool())
+
+	parquetQueryable, err := search.NewParquetQueryable(cDecoder, func(ctx context.Context, mint, maxt int64) ([]*parquet_storage.ParquetShard, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		blocks, ok := ExtractBlocksFromContext(ctx)
+		if !ok {
+			return nil, errors.Errorf("failed to extract blocks from context")
+		}
+		userBkt := bucket.NewUserBucketClient(userID, bucketClient, limits)
+
+		shards := make([]*parquet_storage.ParquetShard, 0, len(blocks))
+
+		for _, block := range blocks {
+			// we always only have 1 shard - shard 0
+			shard, err := parquet_storage.OpenParquetShard(ctx, userBkt, block.ID.String(), 0)
+			if err != nil {
+				return nil, err
+			}
+			shards = append(shards, shard)
+		}
+
+		return shards, nil
+	})
+
+	p := &parquetQueryableWithFallback{
+		subservices:           manager,
+		blockStorageQueryable: blockStorageQueryable,
+		parquetQueryable:      parquetQueryable,
+		queryStoreAfter:       config.QueryStoreAfter,
+		subservicesWatcher:    services.NewFailureWatcher(),
+		finder:                blockStorageQueryable.finder,
+		metrics:               newParquetQueryableFallbackMetrics(reg),
+	}
+
+	p.Service = services.NewBasicService(p.starting, p.running, p.stopping)
+
+	return p, err
+}
+
+func (p *parquetQueryableWithFallback) starting(ctx context.Context) error {
+	p.subservicesWatcher.WatchManager(p.subservices)
+	if err := services.StartManagerAndAwaitHealthy(ctx, p.subservices); err != nil {
+		return errors.Wrap(err, "unable to start blocks storage queryable subservices")
+	}
+	return nil
+}
+
+func (p *parquetQueryableWithFallback) running(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-p.subservicesWatcher.Chan():
+			return errors.Wrap(err, "block storage queryable subservice failed")
+		}
+	}
+}
+
+func (p *parquetQueryableWithFallback) stopping(_ error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), p.subservices)
+}
+
+func (p *parquetQueryableWithFallback) Querier(mint, maxt int64) (storage.Querier, error) {
+	pq, err := p.parquetQueryable.Querier(mint, maxt)
+	if err != nil {
+		return nil, err
+	}
+
+	bsq, err := p.blockStorageQueryable.Querier(mint, maxt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &parquetQuerier{
+		minT:               mint,
+		maxT:               maxt,
+		parquetQuerier:     pq,
+		queryStoreAfter:    p.queryStoreAfter,
+		blocksStoreQuerier: bsq,
+		finder:             p.finder,
+		metrics:            p.metrics,
+	}, nil
+}
+
+type parquetQuerier struct {
+	minT, maxT int64
+
+	finder BlocksFinder
+
+	parquetQuerier     storage.Querier
+	blocksStoreQuerier storage.Querier
```

Comment:
It is weird to have a parquetQuerier inside a parquetQuerier. Should we change the struct name to `parquetQuerierWithFallback` if it actually has 2 queriers?

Reply:
indeed! i got confused myself! Will change it!